### PR TITLE
ServerConnectionAcceptor for simplifying setup vs. resume scenario

### PR DESCRIPTION
**Pull request by lehecka**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -180,7 +180,7 @@ add_library(
   src/FrameTransport.cpp
   src/FrameTransport.h
   src/FrameProcessor.h
-  src/ReactiveSocket.h src/StreamsFactory.cpp src/StreamsFactory.h)
+  src/ReactiveSocket.h src/StreamsFactory.cpp src/StreamsFactory.h src/ServerConnectionAcceptor.cpp src/ServerConnectionAcceptor.h)
```
Comment (yschimke) [inline CMakeLists.txt]: can you reformat these, or combine the rest onto fewer lines?

File: src/ConnectionSetupPayload.cpp [hunk 2]
```diff
@@ -9,6 +9,8 @@ std::ostream& operator<<(
     const ConnectionSetupPayload& setupPayload) {
   return os << "[metadataMimeType: " << setupPayload.metadataMimeType
             << " dataMimeType: " << setupPayload.dataMimeType
-            << " payload: " << setupPayload.payload << "]";
+            << " payload: " << setupPayload.payload << "]"
```
Comment (yschimke) [inline src/ConnectionSetupPayload.cpp]: why end the bracket here?

File: src/ConnectionSetupPayload.cpp [hunk 3]
```diff
@@ -9,6 +9,8 @@ std::ostream& operator<<(
...
+            << " token: " << setupPayload.token.toString()
```
Comment (yschimke) [inline src/ConnectionSetupPayload.cpp]: should token support operator<<?

File: test/simple/StatsPrinter.h [hunk 4]
```diff
@@ -24,5 +24,8 @@ class StatsPrinter : public Stats {
   void frameWritten(const std::string& frameType) override;
   void frameRead(const std::string& frameType) override;
   void resumeBufferChanged(int framesCountDelta, int dataSizeDelta) override;
+
+  void connectionClosedInServerConnectionAcceptor(
```
Comment (yschimke) [inline test/simple/StatsPrinter.h]: nit: this seems awfully specific.  Stats is really only for reporting on errors, could we use a more generic name.  badIncomingConnection?

File: test/resume/TcpResumeServer.cpp [hunk 5]
```diff
@@ -175,35 +142,23 @@ class ServerRequestHandler : public DefaultRequestHandler {
...
-        eventBase_(eventBase),
-        stats_(stats){};
+  MyServerConnectionAcceptor(EventBase& eventBase, Stats& stats)
+      : ServerConnectionAcceptor(stats), eventBase_(eventBase) {}
 
-  virtual ~Callback() = default;
+  void setupNewSocket(
+      std::shared_ptr<FrameTransport> frameTransport,
+      ConnectionSetupPayload setupPayload) override {
+    CHECK(setupPayload.resumable);
```
Comment (yschimke) [inline test/resume/TcpResumeServer.cpp]: it seems useful that this server is written as an example that can accept requests from either the resume client or standard client and just fallback to no resumption.  e.g. CHECK seems extreme.

File: test/resume/TcpResumeServer.cpp [hunk 6]
```diff
@@ -223,26 +178,57 @@ class Callback : public AsyncServerSocket::AcceptCallback {
...
-                    ResumeIdentificationToken>& kv) {
-            return kv.first.get() == &socket;
-          }));
-    }
+  void resumeSocket(
+      std::shared_ptr<FrameTransport> frameTransport,
+      ResumeIdentificationToken token,
+      ResumePosition position) override {
+    LOG(INFO) << "MyServerConnectionAcceptor::resumeSocket resume token "
+              << token.toString();
```
Comment (yschimke) [inline test/resume/TcpResumeServer.cpp]: what are we, java developers?

**Comment by yschimke** [comment]
LGTM - I'd open it up for wider discussion though

File: src/ConnectionSetupPayload.cpp (hunk 2 above)
Comment (lehecka) [inline src/ConnectionSetupPayload.cpp]: I will remove them completely..

File: src/ConnectionSetupPayload.cpp (hunk 3 above)
Comment (lehecka) [inline src/ConnectionSetupPayload.cpp]: we couldn't agree on the format so we didn't make it printable thru <<

File: test/simple/StatsPrinter.h (hunk 4 above)
Comment (lehecka) [inline test/simple/StatsPrinter.h]: I think this should be very specific because it is used for one specific case. I believe we should have way more specific counters all over the library. Essentially the idea is that for debugging reasons you want to know if you library is hitting any edge cases in production, if you are refusing connections / closing / rejecting frames and such. Would you agree with that statement?

File: test/resume/TcpResumeServer.cpp (hunk 5 above)
Comment (lehecka) [inline test/resume/TcpResumeServer.cpp]: since this is a test code, I dont care. Will remove the check

File: test/resume/TcpResumeServer.cpp (hunk 6 above)
Comment (lehecka) [inline test/resume/TcpResumeServer.cpp]: :) 
I guess not.
